src/v/raft/id_allocator_stm.cc

```diff
@@ -0,0 +1,249 @@
+    b.add_raw_kv(std::move(key_buf), std::move(v_buf));
+    return std::move(b).build();
+}
+
+id_allocator_stm::id_allocator_stm(
+  ss::logger& logger, raft::consensus* c, config::configuration& config)
+  : raft::state_machine(c, logger, ss::default_priority_class())
+  , _config(config)
+  , _last_seq_tick(0)
+  , _c(c) {}
+
+ss::future<id_allocator_stm::stm_allocation_result>
+id_allocator_stm::allocate_id_and_wait(
+  model::timeout_clock::time_point timeout) {
+    auto prelude = ss::now();
+    auto range = _config.id_allocator_batch_size.value();
+
+    if (_last_allocated_range > 0) {
+        auto allocated_id = _last_allocated_base;
+        _last_allocated_range -= 1;
+        _last_allocated_base += 1;
+
+        return ss::make_ready_future<stm_allocation_result>(
+          stm_allocation_result{allocated_id, raft::errc::success});
+    }
+
+    if (_processed > _config.id_allocator_log_capacity.value()) {
+        sequence_id seq = sequence_id{
+          _run_id.value(), _c->self(), ++_last_seq_tick};
+        prelude = replicate_and_wait(prepare_truncation_cmd{seq}, timeout, seq)
+                    .then([this](bool replicated) {
+                        if (replicated) {
+                            (void)ss::with_gate(_gate, [this] {
+                                return replicate(serialize_cmd(
+                                  execute_truncation_cmd{
+                                    _prepare_offset, _prepare_state},
+                                  id_allocator_stm_batch_type));
+                            });
+                        }
+                    });
+    } else if (_should_cache) {
+        (void)ss::with_gate(_gate, [this] {
+            return replicate(serialize_cmd(
+              execute_truncation_cmd{_prepare_offset, _prepare_state},
+              id_allocator_stm_batch_type));
+        });
+    }
+
+    return prelude.then([this, timeout, range] {
+        sequence_id seq = sequence_id{
+          _run_id.value(), _c->self(), ++_last_seq_tick};
+
+        return replicate_and_wait(allocation_cmd{seq, range}, timeout, seq)
+          .then([this](log_allocation_result r) {
+              _last_allocated_base = r.base + 1;
+              _last_allocated_range = r.range - 1;
+              return stm_allocation_result{r.base, r.raft_status};
+          });
+    });
+}
+
+ss::future<> id_allocator_stm::apply(model::record_batch b) {
+    if (b.header().type == id_allocator_stm::id_allocator_stm_batch_type) {
+        return process(std::move(b));
+    }
+
+    return ss::now();
+}
+
+ss::future<> id_allocator_stm::start() {
+    auto last = _c->read_last_applied();
+    if (last != model::offset{}) {
+        set_next(last);
+    }
+
+    return state_machine::start().then([this] {
+        return _c->get_run_id().then(
+          [this](model::run_id id) { _run_id.emplace(id); });
+    });
+}
+
+ss::future<> id_allocator_stm::process(model::record_batch&& b) {
+    auto r = b.copy_records();
```

Comment:
why do we need records copy ?

Reply:
I followed the pattern we use for other state machines, we use `mux` there and forward the record_batch to `apply_update` of a state machine. Currently there are two of them: `members_manager.h` and `topic_updates_dispatcher.h`. The former uses `copy_records` directly and the latter calls `deserialize` which also uses `copy_records` under the hood.